### PR TITLE
[SPARK-28578][INFRA] Improve Github pull request template

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to improve the Github template for better and faster review iterations and better interactions between PR authors and reviewers.

As suggested in the the [dev mailing list](http://apache-spark-developers-list.1001551.n3.nabble.com/DISCUSS-New-sections-in-Github-Pull-Request-description-template-td27527.html), this PR referred [Kubernates' PR template](https://raw.githubusercontent.com/kubernetes/kubernetes/master/.github/PULL_REQUEST_TEMPLATE.md).

Therefore, those fields are newly added:

```
### Why are the changes needed?
### Does this PR introduce any user-facing change?
```

and some comments were added.

### Why are the changes needed?

Currently, many PR descriptions are poorly formatted, which causes some overheads between PR authors and reviewers.

There are multiple problems by those poorly formatted PR descriptions:

- Some PRs still write single line in PR description with 500+- code changes in a critical path.
- Some PRs do not describe behaviour changes and reviewers need to find and document.
- Some PRs are hard to review without outlines but they are not mentioned sometimes.
- Spark is being old and sometimes we need to track the history deep. Due to poorly formatted PR description,  sometimes it requires to read whole codes of whole commit histories to find the root cause of a bug.
- Reviews take a while but the number of PR still grows.

This PR targets to alleviate the problems and situation.

### Does this PR introduce any user-facing change?

Yes, it changes the PR templates when PRs are open. This PR uses the template this PR proposes.

### How was this patch tested?

Manually tested via Github preview feature.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @apache/spark-committers

**Comment by HyukjinKwon** [comment]
@rxin

File: .github/PULL_REQUEST_TEMPLATE
Comment (dongjoon-hyun) [inline .github/PULL_REQUEST_TEMPLATE]: nit. `http` -> `https`

File: .github/PULL_REQUEST_TEMPLATE
Comment (maropu) [inline .github/PULL_REQUEST_TEMPLATE]: how about adding an example like;
```
3. If the PR is unfinished, add `[WIP]` in your PR title, e.g., [WIP][SPARK-XXXX] Your PR title...
```

Comment (HyukjinKwon) [reply .github/PULL_REQUEST_TEMPLATE]: +1. Let me collect more feedback and address them later in batch.

Comment (maropu) [reply .github/PULL_REQUEST_TEMPLATE]: yea, its ok. thx

Comment (dongjoon-hyun) [inline .github/PULL_REQUEST_TEMPLATE]: In this line, can we explicitly guide more about the PR title, too? Something like the following. (I borrowed this from the contributor guideline because many people missed this.)
> The PR title should be of the form [SPARK-xxxx][COMPONENT] Title, where SPARK-xxxx is the relevant JIRA number

Comment (dongjoon-hyun) [reply .github/PULL_REQUEST_TEMPLATE]: Oh. It will be the same comment with @maropu 's comment.

File: .github/PULL_REQUEST_TEMPLATE
Comment (maropu) [inline .github/PULL_REQUEST_TEMPLATE]: Also, if possible, I'd like PR authors to put a very simple example query in PR descriptions to reproduce an issue (Especially for fixes in the SQL part).

File: .github/PULL_REQUEST_TEMPLATE
Comment (gengliangwang) [inline .github/PULL_REQUEST_TEMPLATE]: In the [Kubernates' PR template](https://raw.githubusercontent.com/kubernetes/kubernetes/master/.github/PULL_REQUEST_TEMPLATE.md), the changes and reason descriptions are put in one section:
```
**What this PR does / why we need it**:
```

Just double-check, do you separate the two parts on purpose? I am actually OK with either way.

Comment (HyukjinKwon) [reply .github/PULL_REQUEST_TEMPLATE]: Few small reasons:
- I feel sure some people will ignore one of both questions often :-).
- Wanted to emphasize the reason when a PR adds an API
-  "What changes were proposed in this pull request?" was already existent - wanted to avoid to change it.

But they are rather minor reasons. I don't mind combining it if you and other people more prefer that way.

File: .github/PULL_REQUEST_TEMPLATE
Comment (srowen) [inline .github/PULL_REQUEST_TEMPLATE]: Nit: "have added or _run_"

**Comment by tgravescs** [comment]
so I think this brings up a good question as to what we want in the jira and what we want in the PR. Many times they have the same information. Do we want it to be copied to both places, referenced, etc.?

**Comment by srowen** [comment]
This is covered in the contributing guide, though that's a big doc. It could be worth a note here that contributors should pay attention to the relationship between JIRAs and PRs as described there. Though I suppose they should really read it all.

File: .github/PULL_REQUEST_TEMPLATE
Comment (HyukjinKwon) [reply .github/PULL_REQUEST_TEMPLATE]: Actually I think that's at `How was this patch tested?`

**Comment by HyukjinKwon** [comment]
I think basically JIRA describes what issue and PR describes how it is fixed.

Each focuses on each's purpose but other information can be duplicated for better and faster reviews. I think it's case-by-case.

**Comment by HyukjinKwon** [comment]
Let me cc some more active contributors who didn't put input here for better visibility and to collect more feedback. cc @viirya @mgaido91 @MaxGekk @HeartSaVioR @gaborgsomogyi @dilipbiswal

**Comment by viirya** [comment]
Thanks.

Don't have stable Internet connection right now.

May look into it and provide feedback tomorrow or later.

File: .github/PULL_REQUEST_TEMPLATE
Comment (MaxGekk) [inline .github/PULL_REQUEST_TEMPLATE]: If only a few test suites check my changes, should I mention those suites here, or I can skip the entire section?

File: .github/PULL_REQUEST_TEMPLATE
Comment (MaxGekk) [inline .github/PULL_REQUEST_TEMPLATE]: Should some DBMSes be mandatory? For example, PostgreSQL and/or Hive?

Comment (HyukjinKwon) [reply .github/PULL_REQUEST_TEMPLATE]: Those are usually what we refer but I didn't mean this as mandatory .. let's avoid to explicitly list up some specific DBMSes..

Comment (HyukjinKwon) [reply .github/PULL_REQUEST_TEMPLATE]: Let me try to clarify.

File: .github/PULL_REQUEST_TEMPLATE
Comment (mgaido91) [inline .github/PULL_REQUEST_TEMPLATE]: I am note sure about this section. I mean, any note can also be left in the other previous sections (e.g a link to a design doc may be put in the `What changes...` and a reference to a mailing list discussion in the `Why changes...`). I feel like this section will we always left empty and the times it'd be filled, we can put the same info in other sections.

File: .github/PULL_REQUEST_TEMPLATE
Comment (mgaido91) [inline .github/PULL_REQUEST_TEMPLATE]: just a curiosity: why are we changing the title level?

**Review by mgaido91 — COMMENTED** [review]
thanks for pinging me @HyukjinKwon

File: .github/PULL_REQUEST_TEMPLATE
Comment (HyukjinKwon) [reply .github/PULL_REQUEST_TEMPLATE]: I think the problem is that all the sections here can be left in `What changes...` or `Why changes...` but people don't do it. I would like to encourage contributors to pay attentions to leave some notes for easier and faster review..

File: .github/PULL_REQUEST_TEMPLATE
Comment (HyukjinKwon) [reply .github/PULL_REQUEST_TEMPLATE]: Just it looked too big to me. I don't mind changing back.

File: .github/PULL_REQUEST_TEMPLATE
Comment (HyukjinKwon) [reply .github/PULL_REQUEST_TEMPLATE]: I think it's just like we did before - let me clarify it.

One thing I would like to emphasize here was to leave a reproducer to show how to test step by step if tests are not added or not enough. Some people still say just "Manually tested" here often.

File: .github/PULL_REQUEST_TEMPLATE
Comment (mgaido91) [reply .github/PULL_REQUEST_TEMPLATE]: no, I was just curious, I am fine whatever :)

File: .github/PULL_REQUEST_TEMPLATE
Comment (mgaido91) [reply .github/PULL_REQUEST_TEMPLATE]: mmmh...then what about giving more suggestions on how to make a review easier to be reviewed instead of creating a dedicated section?

File: .github/PULL_REQUEST_TEMPLATE
Comment (gatorsmile) [inline .github/PULL_REQUEST_TEMPLATE]: minimised -> concise?

File: .github/PULL_REQUEST_TEMPLATE
Comment (gatorsmile) [inline .github/PULL_REQUEST_TEMPLATE]: beteween -> between

File: .github/PULL_REQUEST_TEMPLATE
Comment (gatorsmile) [reply .github/PULL_REQUEST_TEMPLATE]: In most cases, unit tests are not enough. Most PRs should also include end-to-end tests and perf tests. Also, I found most of new PR authors just write the positive cases. We should remind them to add negative test cases too.

File: .github/PULL_REQUEST_TEMPLATE
Comment (gatorsmile) [inline .github/PULL_REQUEST_TEMPLATE]: What we expect in these two sections: `### What changes were proposed in this pull request?` and `### Why changes are needed?`  ? Should we combine them?

Comment (HyukjinKwon) [reply .github/PULL_REQUEST_TEMPLATE]: We can combine but I expected to explicitly mention those two separately:

```
### What changes were proposed in this pull request?

Describe how this PR fixes the JIRA. For instance, A calls B but B has to be C blah blah.

### Why changes are needed?

e.g.,
1. This API has been requested in user mailing list multiple times. See http://... , http:// ...
2. It follows those references to make Spark's behaviour reasonable.
3. This is a bug fix that introduced by SPARK-XXXXX. It should work as expected.
```

Comment (HyukjinKwon) [reply .github/PULL_REQUEST_TEMPLATE]: Because nowadays Spark's being conservative when it adds some APIs and we need to know why such APIs are needed, for instance.

Comment (HyukjinKwon) [reply .github/PULL_REQUEST_TEMPLATE]: and .. they are separate two sections to avoid people ignore one of both questions. @gengliangwang pointed out similar one and I answered at https://github.com/apache/spark/pull/25310#discussion_r309149015.

If you guys prefer combining, I don't mind but I wanted to say why I separated them.

File: .github/PULL_REQUEST_TEMPLATE
Comment (HeartSaVioR) [reply .github/PULL_REQUEST_TEMPLATE]: Once it is merged it doesn't provide any difference so not a big deal.

File: .github/PULL_REQUEST_TEMPLATE
Comment (HeartSaVioR) [reply .github/PULL_REQUEST_TEMPLATE]: +1 on @mgaido91 , and maybe in many cases adding self comments on changeset would be more helpful. I'd rather not expect people don't explain enough on `What changes...` and `Why changes...` but try to explain this section.

File: .github/PULL_REQUEST_TEMPLATE
Comment (HeartSaVioR) [inline .github/PULL_REQUEST_TEMPLATE]: Ideally 3 and 4 should be either JIRA link or another section of PR's content even this section is not introduced.

File: .github/PULL_REQUEST_TEMPLATE
Comment (HeartSaVioR) [reply .github/PULL_REQUEST_TEMPLATE]: Do we add content in your comment to the template? I guess we need to have some explanation on both sections.

Comment (HyukjinKwon) [reply .github/PULL_REQUEST_TEMPLATE]: will clarify it.

File: .github/PULL_REQUEST_TEMPLATE
Comment (HyukjinKwon) [reply .github/PULL_REQUEST_TEMPLATE]: That's right in general, in particular, for SQL test cases but I think it might be different in other components for many reasons. For instance, difficult to write a negative test case, etc. Let me try to clarify it in a general way.

I think we should write things that all people here could generally agree upon for now.

File: .github/PULL_REQUEST_TEMPLATE
Comment (viirya) [inline .github/PULL_REQUEST_TEMPLATE]: How the section shows in the PR desctiption? I means `<!-- ... -->`

File: .github/PULL_REQUEST_TEMPLATE
Comment (viirya) [inline .github/PULL_REQUEST_TEMPLATE]: not sure, about maybe `Why changes are needed?` -> `Why the changes are needed?`

File: .github/PULL_REQUEST_TEMPLATE
Comment (viirya) [inline .github/PULL_REQUEST_TEMPLATE]: Here it mentions `reproducer to test`. Is it overlapped with next section?

File: .github/PULL_REQUEST_TEMPLATE
Comment (cloud-fan) [reply .github/PULL_REQUEST_TEMPLATE]: I think most of the time this section will be empty. If a contributor does have some special notes, he usually will find a way to put it, e.g. at the end of the PR description, or add a new comment.

File: .github/PULL_REQUEST_TEMPLATE
Comment (cloud-fan) [inline .github/PULL_REQUEST_TEMPLATE]: In this case, I'd like to see it in the first section, not after reading the description of the changes and the reasons to change it.

Comment (viirya) [reply .github/PULL_REQUEST_TEMPLATE]: class hierarchy can be seen as part of the change (`What changes were proposed in this pull request`).

File: .github/PULL_REQUEST_TEMPLATE
Comment (felixcheung) [reply .github/PULL_REQUEST_TEMPLATE]: `Why are the changes needed`?

File: .github/PULL_REQUEST_TEMPLATE
Comment (HyukjinKwon) [reply .github/PULL_REQUEST_TEMPLATE]: It's HTML comments so it will be hidden. It will be only seen to who edits the PR description. For instance ... copy and paste this into comment and preview.

```
<!-- ... -->
```

File: .github/PULL_REQUEST_TEMPLATE
Comment (HyukjinKwon) [reply .github/PULL_REQUEST_TEMPLATE]: I think that's slightly different. Unittests vs examples to show the behaviour changes. Unittests can include multiple cases vs the examples should usually be a single example.

Comment (HyukjinKwon) [reply .github/PULL_REQUEST_TEMPLATE]: Let me clarify.

**Comment by HyukjinKwon** [comment]
I believe I address all of the comments and should be ready for a look.

File: .github/PULL_REQUEST_TEMPLATE
Comment (mgaido91) [inline .github/PULL_REQUEST_TEMPLATE]: some nits about wording:
I am not sure about the term `descendants`, but I am not a native speaker, so I hope some native English speaker can shed some light on the best term.
I'd also add that if no UT was added, it'd be great to describe why they weren't added, why it was not possible to do so.

File: .github/PULL_REQUEST_TEMPLATE
Comment (mgaido91) [inline .github/PULL_REQUEST_TEMPLATE]:
```suggestion
  1. If you propose a new API, clarify the use case for a new API.
```

**Review by mgaido91 — COMMENTED** [review]
just few hints about wording, otherwise this template seems fine to me.

File: .github/PULL_REQUEST_TEMPLATE
Comment (HyukjinKwon) [reply .github/PULL_REQUEST_TEMPLATE]: @srowen, does it make sense?

**Comment by HyukjinKwon** [comment]
gentle ping for another look.

File: .github/PULL_REQUEST_TEMPLATE
Comment (gaborgsomogyi) [inline .github/PULL_REQUEST_TEMPLATE]: [Here](https://spark.apache.org/contributing.html) it's not mentioned that the PR title should reflect the solution and not the issue. It's not template related but would be good to mention somewhere because it's a common pattern.

File: .github/PULL_REQUEST_TEMPLATE
Comment (gaborgsomogyi) [inline .github/PULL_REQUEST_TEMPLATE]: This chapter answers the `what and why to solve?` questions but I personally always look for something `how is it solved high level`. If the high level approach is questionable it worth to wait with the deep dive.

File: .github/PULL_REQUEST_TEMPLATE
Comment (gaborgsomogyi) [inline .github/PULL_REQUEST_TEMPLATE]: Just for the sake of my own understanding do we want to hunt for breaking changes here?

**Review by gaborgsomogyi — COMMENTED** [review]
As I've just arrived from vacation doing syncing and reached this. @HyukjinKwon thanks for pinging, added my comments.

File: .github/PULL_REQUEST_TEMPLATE
Comment (BryanCutler) [inline .github/PULL_REQUEST_TEMPLATE]: At first glance this might be a little unclear. I think what you are getting at is that if anything changes in the course of the PR, keep the description updated?
If so then maybe change to "Be sure to keep the PR description updated to reflect all changes"?

File: .github/PULL_REQUEST_TEMPLATE
Comment (BryanCutler) [inline .github/PULL_REQUEST_TEMPLATE]: Is this necessary? There should be a unit test to show the issue, but if not then a concise example should be provided

File: .github/PULL_REQUEST_TEMPLATE
Comment (BryanCutler) [inline .github/PULL_REQUEST_TEMPLATE]: I would remove `Therefore`.  Typo: `write` -> `writing`.  Also remove `some`

File: .github/PULL_REQUEST_TEMPLATE
Comment (BryanCutler) [inline .github/PULL_REQUEST_TEMPLATE]: `leave the link` -> `add the link`

File: .github/PULL_REQUEST_TEMPLATE
Comment (BryanCutler) [inline .github/PULL_REQUEST_TEMPLATE]: typo: `fixes` -> `fix`

File: .github/PULL_REQUEST_TEMPLATE
Comment (BryanCutler) [inline .github/PULL_REQUEST_TEMPLATE]: remove `some`

File: .github/PULL_REQUEST_TEMPLATE
Comment (BryanCutler) [inline .github/PULL_REQUEST_TEMPLATE]: `can test and check through` -> `can test and check`

**Review by BryanCutler — COMMENTED** [review]
Looks great, thanks @HyukjinKwon !  I just had some minor grammatical suggestions.

File: .github/PULL_REQUEST_TEMPLATE
Comment (HyukjinKwon) [reply .github/PULL_REQUEST_TEMPLATE]: Basically yea. For instance, if we meet a breaking change after upgrading to higher Spark, time to time we should look through the commit history. I think it's better to clarify it if that can be identified early.

Comment (gaborgsomogyi) [reply .github/PULL_REQUEST_TEMPLATE]: OK, thanks for the explanation.

**Comment by HyukjinKwon** [comment]
I was stuck in some works. Let me update it tomorrow in KST.

File: .github/PULL_REQUEST_TEMPLATE
Comment (younggyuchun) [reply .github/PULL_REQUEST_TEMPLATE]: typo: calrify -> clarify

Comment (younggyuchun) [inline .github/PULL_REQUEST_TEMPLATE]: typo: calrify -> clarify

File: .github/PULL_REQUEST_TEMPLATE
Comment (younggyuchun) [inline .github/PULL_REQUEST_TEMPLATE]: If there is a design documentation -> If there is design documentation.

File: .github/PULL_REQUEST_TEMPLATE
Comment (viirya) [inline .github/PULL_REQUEST_TEMPLATE]: nit: not big deal, but there are `behavior` and `behaviour`. should we use only one?

Comment (srowen) [reply .github/PULL_REQUEST_TEMPLATE]: I would stick with US spelling in general (behavior)

Comment (HyukjinKwon) [reply .github/PULL_REQUEST_TEMPLATE]: Ah, I'm used to UK spelling. Im trying to stick to US but keeps missing :(. Let me fix.

File: .github/PULL_REQUEST_TEMPLATE
Comment (HyukjinKwon) [reply .github/PULL_REQUEST_TEMPLATE]: Nope, it's not necessary but just for a faster review. None of them here are strictly required but rather guidelines.

File: .github/PULL_REQUEST_TEMPLATE
Comment (HyukjinKwon) [reply .github/PULL_REQUEST_TEMPLATE]: I think useful notes cover this high level stuff:

```
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
```

It's not always to know high level approach I guess so it should be fine.

**Comment by HyukjinKwon** [comment]
I address the comments. Should be ready for a look.

**Comment by gaborgsomogyi** [comment]
retest this please

**Review by gaborgsomogyi — APPROVED** [review]

**Review by BryanCutler — APPROVED** [review]
LGTM

**Comment by HyukjinKwon** [comment]
Let me merge this one in few days if there are no more comments.

**Review by cloud-fan — APPROVED** [review]

**Review by gengliangwang — APPROVED** [review]

**Review by MaxGekk — APPROVED** [review]

**Review by HeartSaVioR — APPROVED** [review]
LGTM

**Review by viirya — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HyukjinKwon** [comment]
Let's see how it goes. thanks guys!

**Pull request closed without merging**
